### PR TITLE
Wallet staking add validors hook

### PR DESCRIPTION
**Pull request by Jibz1**
- Added useGetObject, useValidators and useGetNFTMetaById
- Added useInvalidateValidatorsCache to invalidate cache after a staking
- added validator response type (this types should come from the SDK)

### CODE REVIEW

File: apps/wallet/src/ui/app/hooks/useGetNFTMetaById.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+import { useGetObject } from './useGetObject';
+
+export type NFTMetadata = {
+    name: string | null;
+    description: string | null;
+    url: string;
+};
+
+export function useGetNFTMetaById(objectID: string): NFTMetadata | null {
```
Comment (pchrysochoidis) [inline apps/wallet/src/ui/app/hooks/useGetNFTMetaById.ts:15]: remove already on main with different name

File: apps/wallet/src/ui/app/hooks/useGetObject.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+import {
+    useQuery,
+    type UseQueryResult,
+    useQueryClient,
+} from '@tanstack/react-query';
+
+//TODO use hook useRpc -
+import { api } from '../redux/store/thunk-extras';
+
+export function useGetObject(
```
Comment (pchrysochoidis) [inline apps/wallet/src/ui/app/hooks/useGetObject.ts:14]: remove this already in main

File: apps/wallet/src/ui/app/hooks/useGetObject.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+            return rpc.getObject(objectId);
+        },
+        { enabled: !!objectId }
+    );
+
+    return response;
+}
+
+// Invalidates the cache Objects. called after a transaction
+export function useValidateObjectQueryCache(cacheId: string) {
```
Comment (pchrysochoidis) [inline apps/wallet/src/ui/app/hooks/useGetObject.ts:30]: not sure about this, I think we should use mutations and as a side effect invalidate cache. This one seems that it will invalidate cache in every render

File: apps/wallet/src/ui/app/staking/home/ValidatorDataTypes.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//TODO pull from the SDK
```
Comment (pchrysochoidis) [inline apps/wallet/src/ui/app/staking/home/ValidatorDataTypes.ts:4]: I think Jordan has added similar types to SDK, can we use those?

Comment (Jibz1) [reply apps/wallet/src/ui/app/staking/home/ValidatorDataTypes.ts:4]: I think so, but I am not sure if it has landed

File: apps/wallet/src/ui/app/hooks/useGetObject.ts (hunk 3 above)
Comment (Jibz1) [reply apps/wallet/src/ui/app/hooks/useGetObject.ts:30]: Yeah since it is a hook, I am not sure how mutation will work. I want something on demand. eg after a successful transaction/staking

Comment (pchrysochoidis) [reply apps/wallet/src/ui/app/hooks/useGetObject.ts:29]: We should use a mutation to do the transaction/staking and there in the mutation onSuccess invalidate the cache.

Comment (Jibz1) [reply apps/wallet/src/ui/app/hooks/useGetObject.ts:29]: Got it

File: apps/wallet/src/ui/app/hooks/index.ts [hunk 5]
```diff
@@ -16,9 +16,11 @@ export { default as useNFTBasicData } from './useNFTBasicData';
 export { useObjectsState } from './useObjectsState';
 export { useWaitForElement } from './useWaitForElement';
 export { useFormatCoin, useCoinDecimals } from './useFormatCoin';
-export { useGetNFTMeta } from './useGetNFTMeta';
+export { useGetNFTMetaById } from './useGetNFTMetaById';
```
Comment (Jordan-Mysten) [inline apps/wallet/src/ui/app/hooks/index.ts:19]: Curious why the rename?

File: apps/wallet/src/ui/app/hooks/useGetValidators.ts [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+import { useGetObject } from './useGetObject';
+
+//TODO: Remove when available on SDK, types should come from the SDK
+import type {
+    Validator,
+    ValidatorState,
+} from '_app/staking/home/ValidatorDataTypes';
+
+export const VALIDATORS_OBJECT_ID = '0x05';
```
Comment (Jordan-Mysten) [inline apps/wallet/src/ui/app/hooks/useGetValidators.ts:16]: Can we either move this to the SDK or share it internally within the app (e.g. pending delegations)?

File: apps/wallet/src/ui/app/hooks/useGetObject.ts [hunk 7]
```diff
@@ -6,7 +6,7 @@ import { useQuery, type UseQueryResult } from '@tanstack/react-query';
 
 import { useRpc } from '_hooks';
 
-export function useGetObjectData(
+export function useGetObject(
     objectId: string
```
Comment (Jordan-Mysten) [inline apps/wallet/src/ui/app/hooks/useGetObject.ts:10]: Something I just thought of here: We should normalize the object ID when we fetch it so that we can have the most cache hits possible (e.g. `0x2` and `0x0000000000000000000000000000000000000002` should be cached under the same entry).

We can do this later though, added https://mysten.atlassian.net/browse/APPS-299 to track for now.

File: apps/wallet/src/ui/app/hooks/useGetValidators.ts [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+function processValidators(
+    set: Validator[],
+    totalStake: bigint,
+    current_epoch: number,
+    walletAddress?: string | null
+) {
+    return set.map((av) => {
+        const rawName = av.fields.metadata.fields.name;
+
+        const name = Buffer.from(rawName, 'base64').toString();
```
Comment (Jordan-Mysten) [inline apps/wallet/src/ui/app/hooks/useGetValidators.ts:27]: There's new name processing logic that recently landed that you need to follow (it's annoying while we transition some data types), you can see it here: https://github.com/MystenLabs/sui/blob/main/apps/wallet/src/ui/app/staking/usePendingDelegation.tsx#L90-L99

File: apps/wallet/src/ui/app/hooks/useGetValidators.ts [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+//TODO: Remove when available on SDK, types should come from the SDK
+import type {
+    Validator,
+    ValidatorState,
+} from '_app/staking/home/ValidatorDataTypes';
+
+export const VALIDATORS_OBJECT_ID = '0x05';
+
+function processValidators(
+    set: Validator[],
```
Comment (Jordan-Mysten) [inline apps/wallet/src/ui/app/hooks/useGetValidators.ts:19]: I'd call this `validators`

File: apps/wallet/src/ui/app/hooks/useGetValidators.ts [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+        data && isSuiObject(data.details) && isSuiMoveObject(data.details.data)
+            ? (data.details.data.fields as ValidatorState)
+            : null;
+
+    const totalStake =
+        validatorsData?.validators.fields.total_validator_stake || 0n;
+
+    const validators = useMemo(() => {
+        if (!validatorsData) return [];
+        const processedValidators = processValidators(
```
Comment (Jordan-Mysten) [inline apps/wallet/src/ui/app/hooks/useGetValidators.ts:96]: I'd inline `processValidators` here, does't seem like we need a separate function here.

File: apps/wallet/src/ui/app/hooks/useGetValidators.ts [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+            (365 / num_epochs_participated - 1);
+
+        const pending_delegationsByAddress = pending_delegations
+            ? pending_delegations.filter(
+                  (d) => d.fields.delegator === walletAddress
+              )
+            : [];
+
+        return {
+            name: name,
```
Comment (Jordan-Mysten) [inline apps/wallet/src/ui/app/hooks/useGetValidators.ts:49]: I get __some__ of this extraction (e.g. `name` where it needs custom processing), but a lot of this seems like you could just return the entire `av` object, and do a lot of this processing where you need it in the UI. I'm very unconvinced that this hook really should exist. The pending delegation is a good example of something that I think is good to be a custom hook because it's a bit complex and also it's also not even generalized, it's just a helper hook within the staking code.

I'd probably recommend doing a lot more of this computation within the UI, and just making this return either the entire state object, or the `active_validators` with some light metadata.

File: apps/wallet/src/ui/app/hooks/useGetValidators.ts [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+                0n
+            ),
+            metadata: av.fields.metadata,
+            // TODO: update
+            pendingDelegationsCount: pending_delegations.length,
+            totalPendingDelegationAmount: BigInt(
+                av.fields.metadata.fields.next_epoch_delegation || 0n
+            ),
+            logo: null,
+            suiEarned: 0n,
```
Comment (Jordan-Mysten) [inline apps/wallet/src/ui/app/hooks/useGetValidators.ts:72]: Why is this 0?

File: apps/wallet/src/ui/app/staking/home/ValidatorDataTypes.ts (hunk 4 above)
Comment (Jordan-Mysten) [reply apps/wallet/src/ui/app/staking/home/ValidatorDataTypes.ts:4]: It hasn't landed, I'm trying to land a much larger schema change first, will probably land it later.

**Review by Jordan-Mysten — CHANGES_REQUESTED** [review]

File: apps/wallet/src/ui/app/hooks/index.ts (hunk 5 above)
Comment (Jibz1) [reply apps/wallet/src/ui/app/hooks/index.ts]: revert this back

File: apps/wallet/src/ui/app/hooks/useGetObject.ts (hunk 7 above)
Comment (Jibz1) [reply apps/wallet/src/ui/app/hooks/useGetObject.ts:13]: done

File: apps/wallet/src/ui/app/hooks/useGetValidators.ts (hunk 8 above)
Comment (Jibz1) [reply apps/wallet/src/ui/app/hooks/useGetValidators.ts]: done

File: apps/wallet/src/ui/app/hooks/useGetValidators.ts (hunk 9 above)
Comment (Jibz1) [reply apps/wallet/src/ui/app/hooks/useGetValidators.ts]: remove processValidators

File: apps/wallet/src/ui/app/hooks/useGetValidators.ts (hunk 12 above)
Comment (Jibz1) [reply apps/wallet/src/ui/app/hooks/useGetValidators.ts]: I had this temp until sui earned was fully flush out.

File: apps/wallet/src/ui/app/staking/usePendingDelegation.tsx [hunk 13]
```diff
@@ -9,8 +9,8 @@ import { notEmpty } from '../helpers';
 import { useAppSelector } from '../hooks';
 import { api } from '../redux/store/thunk-extras';
 
-const STATE_OBJECT = '0x5';
-const VALDIATOR_NAME = /^[A-Z-_.\s0-9]+$/i;
+export const STATE_OBJECT = '0x5';
+export const VALDIATOR_NAME = /^[A-Z-_.\s0-9]+$/i;
```
Comment (Jordan-Mysten) [inline apps/wallet/src/ui/app/staking/usePendingDelegation.tsx:13]: Maybe just generalize some `getName` function here too to de-dupe the name logic?

File: apps/wallet/src/ui/app/hooks/useGetValidators.ts [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+            const {
+                sui_balance,
+                starting_epoch,
+                pending_delegations,
+                delegation_token_supply,
+            } = av.fields.delegation_staking_pool.fields;
+
+            const num_epochs_participated =
+                validatorsData.epoch - starting_epoch;
+            const APY =
```
Comment (Jordan-Mysten) [inline apps/wallet/src/ui/app/hooks/useGetValidators.ts:58]: Did you verify this formula? I assumed that the `^` in the pseudocode was exponent, but in JS this is bitwise XOR so wanted to make sure this is correct.

File: apps/wallet/src/ui/app/hooks/useGetValidators.ts [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+                        delegation_token_supply.fields.value) ^
+                (365 / num_epochs_participated - 1);
+
+            const pending_delegationsByAddress = pending_delegations
+                ? pending_delegations.filter(
+                      (d) => d.fields.delegator === walletAddress
+                  )
+                : [];
+
+            return {
```
Comment (Jordan-Mysten) [inline apps/wallet/src/ui/app/hooks/useGetValidators.ts:70]: I still remain unconvinced that this is a hook that needs to exist.

File: apps/wallet/src/ui/app/hooks/useGetValidators.ts (hunk 14 above)
Comment (Jibz1) [reply apps/wallet/src/ui/app/hooks/useGetValidators.ts:58]: You are right. totally missed that.

File: apps/wallet/src/ui/app/hooks/useGetValidators.ts (hunk 15 above)
Comment (Jibz1) [reply apps/wallet/src/ui/app/hooks/useGetValidators.ts:70]: We use most of this base meta information across multiple components. Do you mean I repeat the APY calculation across components?

File: apps/wallet/src/ui/app/staking/usePendingDelegation.tsx (hunk 13 above)
Comment (Jibz1) [reply apps/wallet/src/ui/app/staking/usePendingDelegation.tsx:13]: done

File: apps/wallet/src/ui/app/hooks/useGetValidators.ts (hunk 14 above)
Comment (Jibz1) [reply apps/wallet/src/ui/app/hooks/useGetValidators.ts]: done

File: apps/wallet/src/ui/app/hooks/useGetValidators.ts (hunk 15 above)
Comment (Jordan-Mysten) [reply apps/wallet/src/ui/app/hooks/useGetValidators.ts]: Yeah, or just have this hook do the plain lookup of validators in the system object. Otherwise it seems better suited to move into the display logic. Otherwise these ends up being effectively view-models, which we don't want. We want to pass around the raw underlying data when we can because it makes it a lot easier to build out.

**Review by Jordan-Mysten — APPROVED** [review]
